### PR TITLE
Refactor MzMLWriter: Simplify implementation and improve CV compliance

### PR DESCRIPTION
**Pull request by mlorenz49**
## Overview
Major refactoring of the MzMLWriter class to create a cleaner, more maintainable implementation with improved PSI-MS controlled vocabulary compliance. Note that this PR is supposed to be a replacement of #98.

### 🔧 **Core Improvements**
- Remove indexing functionality and multiprocessing for simplified base implementation
- Add comprehensive type annotations and detailed docstrings throughout
- Create `cv_constants.py` module with all PSI-MS controlled vocabulary terms
- Consolidate binary array creation logic and improve error handling

### ✅ **CV Compliance Fixes**
- Ensure all CV terms reference official PSI-MS controlled vocabulary

### 📝 **Code Quality**
- Extract all CV constants to dedicated module for maintainability
- Add comprehensive type hints for better IDE support and static analysis
- Improve parameter documentation with detailed docstrings
- Better separation of concerns and cleaner method organization

### 🧪 **Testing**
- Update test suite with CV compliance validation
- Add tests for binary precision, compression, and integer formatting
- Test edge cases including empty spectra handling
- Validate XML structure and required mzML elements

## Breaking Changes
- ⚠️ Remove `indexed` parameter from `save_mzml` method
- ⚠️ Remove multiprocessing options for cleaner interface

## **Compatibility:**
- mzML files work when importing into Skyline and searching with MSFragger
- these mzML files can be searched with DIANN after conversion to indexed mzML
with MSConvert

### CODE REVIEW

**Comment by jalew188** [comment]
@mlorenz49  Do we need mzml reader more than the writer? I mean, what is the use senario of this writer?

File: tests/unit/test_mzml_writter.py [hunk 1]
```diff
@@ -0,0 +1,372 @@
...
+        start, end = self.spectrum_df[["peak_start_idx", "peak_stop_idx"]].values[
+            spec_idx, :
+        ]
+        return (
+            self.peak_df.mz.values[start:end],
+            self.peak_df.intensity.values[start:end],
+        )
+
+
+class TestMzMLWriter(unittest.TestCase):
```
Comment (mschwoer) [inline tests/unit/test_mzml_writter.py:75]: isn't the rest of the package using `pytest`? if so, please adapt for consistency

File: tests/unit/test_mzml_writter.py [hunk 2]
```diff
@@ -0,0 +1,372 @@
...
+
+        tree = ET.parse(output_path)
+        root = tree.getroot()
+        ns = {"mzml": writer.ns_uri}
+
+        # Check MS level values are integers
+        ms_level_params = root.findall(".//mzml:cvParam[@accession='MS:1000511']", ns)
+        for param in ms_level_params:
+            value = param.get("value")
+            assert "." not in value, f"MS level should be integer, got {value}"
```
Comment (mschwoer) [inline tests/unit/test_mzml_writter.py:366]: this check for integers seems a bit hacky to me ;-)

File: alpharaw/mzml_io/mzml_writer.py [hunk 3]
```diff
@@ -0,0 +1,737 @@
...
+        ValueError
+            If binary_precision is not 32 or 64
+            If compression is not None or 'zlib'
+        """
+        if binary_precision not in [32, 64]:
+            raise ValueError("binary_precision must be 32 or 64")
+        if compression not in [None, "zlib"]:
+            raise ValueError("compression must be None or 'zlib'")
+
+        self.ms_data = ms_data
```
Comment (mschwoer) [inline alpharaw/mzml_io/mzml_writer.py:75]: could these be private?

File: alpharaw/mzml_io/mzml_writer.py [hunk 4]
```diff
@@ -0,0 +1,737 @@
...
+        cv_ref: str,
+        accession: str,
+        name: str,
+        value: Union[str, int, float] = "",
+        unit_cv_ref: Optional[str] = None,
+        unit_accession: Optional[str] = None,
+        unit_name: Optional[str] = None,
+    ) -> None:
+        """
+        Helper method to add a CV parameter to an XML element.
```
Comment (mschwoer) [inline alpharaw/mzml_io/mzml_writer.py:691]: maybe define CV here?

**Comment by mlorenz49** [comment]
> @mlorenz49 Do we need mzml reader more than the writer? I mean, what is the use senario of this writer?

The use case is when we manipulate/filter raw files, e.g. when we're working with hybridDIA data but want to use common tools from the community for down stream processing. In this case, we can't rely solely on an .hdf file format because most software doesn't support this format. mzML files are a good solution here because they are the community standard which most tools can work with and there is a consens on how these files should look like (https://www.psidev.info/mzml).

**Comment by jalew188** [comment]
> > @mlorenz49 Do we need mzml reader more than the writer? I mean, what is the use senario of this writer?
> 
> The use case is when we manipulate/filter raw files, e.g. when we're working with hybridDIA data but want to use common tools from the community for down stream processing. In this case, we can't rely solely on an .hdf file format because most software doesn't support this format. mzML files are a good solution here because they are the community standard which most tools can work with and there is a consens on how these files should look like (https://www.psidev.info/mzml).

I see, thank you for quick response.

File: tests/unit/test_mzml_writter.py [hunk 5]
```diff
@@ -0,0 +1,372 @@
...
+
+        # Check that the file was created
+        assert os.path.exists(output_path)
+
+        # Try to parse the file to make sure it's valid XML
+        try:
+            tree = ET.parse(output_path)
+            root = tree.getroot()
+        except Exception as e:
+            pytest.fail(f"Failed to parse the generated mzML file: {e}")
```
[the remark this answers is not in the document]
Comment (mlorenz49) [reply tests/unit/test_mzml_writter.py:101]: It is good to test if the generated mzML file was correctly written and can be read in again, no? Alternatively, we could also try using the mzML reader in order to reimport created files and check if everything worked out well.

Or do you specifically mean that raising the exception is unnecessary?

[the remark this answers is not in the document]
Comment (mschwoer) [reply tests/unit/test_mzml_writter.py:101]: sorry, this was a bit slim in terms of context .. yes, I was only referring to the re-raising .. if a test raises outside of a `with pytest.raises(): ..` block then it fails anyway, and as the tests is run (and fixed ;-)) by a developer anyway, the "pure" error message totally suffices

File: requirements/requirements.txt [hunk 6]
```diff
@@ -10,6 +10,7 @@ pandas==2.2.2
 numpy==1.26.4
```
Comment (mschwoer) [inline requirements/requirements.txt:10]: on [psi-ms.owl](https://github.com/MannLabs/alpharaw/pull/100/files#diff-2837f6b948cc21ba402dff91f52f2686f5289363797b08c95f1c3678db787d48):

move to a `resources` directory, add a readme file which indicates the origin and version (e.g. commit hash) of it ..

**Review by mschwoer — CHANGES_REQUESTED** [review]

**Comment by mlorenz49** [comment]
@mschwoer we might need to update the writer in the future, as I've seen that there are some detailed infos which some engines look for.

E.g. msfragger was looking for some information if FAIMS was used. There is a psi ID for this MS:1001581, however we would need to extract the info with alpharaw first, which again would require us to change the MsFilereader code in order to get that info. Nonetheless, Msfragger runs through.

I would consider this a very vanilla form of the writer which contains the most crucial information required by most software to process.
